The PostCommit SQL job is flaky

The PostCommit SQL is failing over 50% of the time.
Please visit https://github.com/apache/beam/actions/workflows/beam_PostCommit_SQL.yml?query=is%3Afailure+branch%3Amaster to see all failed workflow runs.
See also Grafana statistics: http://metrics.beam.apache.org/d/CTYdoxP4z/ga-post-commits-status?orgId=1&viewPanel=11&var-Workflow=PostCommit%20SQL

`PubsubToIcebergIT > testSimpleInsertFlat FAILED
    org.apache.iceberg.exceptions.RuntimeIOException at PubsubToIcebergIT.java:128
        Caused by: org.apache.hadoop.fs.UnsupportedFileSystemException at PubsubToIcebergIT.java:128`

--- a/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithKeys.java
+++ b/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithKeys.java
@@ -110,33 +110,27 @@ public class WithKeys<K, V> extends PTransform<PCollection<V>, PCollection<KV<K,
 
   @Override
   public PCollection<KV<K, V>> expand(PCollection<V> in) {
-    SerializableFunction<V, K> localFn = fn;
-    TypeDescriptor<V> inputType = in.getTypeDescriptor();
-    TypeDescriptor<KV<K, V>> outputType = getOutputTypeDescriptor(inputType);
     PCollection<KV<K, V>> result =
         in.apply(
             "AddKeys",
-            outputType == null
-                ? MapElements.via(
-                    new SimpleFunction<V, KV<K, V>>() {
-                      @Override
-                      public KV<K, V> apply(V element) {
-                        return KV.of(localFn.apply(element), element);
-                      }
-                    })
-                : MapElements.into(outputType)
-                    .via(
-                        (SerializableFunction<V, KV<K, V>>)
-                            element -> KV.of(localFn.apply(element), element)));
+            MapElements.via(
+                new SimpleFunction<V, KV<K, V>>() {
+                  @Override
+                  public KV<K, V> apply(V element) {
+                    return KV.of(fn.apply(element), element);
+                  }
+                }));
 
     try {
+      Coder<K> keyCoder;
       CoderRegistry coderRegistry = in.getPipeline().getCoderRegistry();
-      if (outputType == null) {
-        Coder<K> keyCoder = coderRegistry.getOutputCoder(fn, in.getCoder());
-        result.setCoder(KvCoder.of(keyCoder, in.getCoder()));
+      if (keyType == null) {
+        keyCoder = coderRegistry.getOutputCoder(fn, in.getCoder());
       } else {
-        result.setCoder(coderRegistry.getCoder(outputType, checkNotNull(inputType), in.getCoder()));
+        keyCoder = coderRegistry.getCoder(keyType);
       }
+      // TODO: Remove when we can set the coder inference context.
+      result.setCoder(KvCoder.of(keyCoder, in.getCoder()));
     } catch (CannotProvideCoderException exc) {
       if (keyType != null) {
         try {
@@ -156,13 +150,5 @@ public class WithKeys<K, V> extends PTransform<PCollection<V>, PCollection<KV<K,
     }
 
     return result;
-  }
-
-  private @Nullable TypeDescriptor<KV<K, V>> getOutputTypeDescriptor(
-      @Nullable TypeDescriptor<V> inputType) {
-    if (keyType == null || inputType == null) {
-      return null;
-    }
-    return TypeDescriptors.kvs(keyType, inputType);
   }
 }

--- a/sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/WithKeysTest.java
+++ b/sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/WithKeysTest.java
@@ -173,21 +173,6 @@ public class WithKeysTest {
   }
 
   @Test
-  public void withKeyTypeShouldSetOutputTypeDescriptorFromInputType() {
-    PCollection<String> values =
-        p.apply(Create.of("1234", "3210").withType(TypeDescriptors.strings()));
-
-    PCollection<KV<Integer, String>> kvs =
-        values.apply(
-            WithKeys.of((SerializableFunction<String, Integer>) Integer::valueOf)
-                .withKeyType(TypeDescriptors.integers()));
-
-    assertEquals(
-        TypeDescriptors.kvs(TypeDescriptors.integers(), TypeDescriptors.strings()),
-        kvs.getTypeDescriptor());
-  }
-
-  @Test
   @Category(NeedsRunner.class)
   public void withLambdaAndNoTypeDescriptorShouldThrow() {
 

